pkg/util/helper/workstatus.go

```diff
@@ -36,16 +37,16 @@ const (
 
 // AggregateResourceBindingWorkStatus will collect all work statuses with current ResourceBinding objects,
 // then aggregate status info to current ResourceBinding status.
-func AggregateResourceBindingWorkStatus(c client.Client, binding *workv1alpha2.ResourceBinding, workload *unstructured.Unstructured) error {
+func AggregateResourceBindingWorkStatus(c client.Client, binding *workv1alpha2.ResourceBinding, workload *unstructured.Unstructured) (controllerutil.OperationResult, error) {
```

Comment:
I think the `OperationResult` is dedicated to presenting the result of a `CreateOrUpdate call`, I'm afraid it would lead to confusing. 
Let me see how we can improve it here.

Reply:
At present, the main distinction is the following three states, one is that the update fails, the second is that the state has not changed, and finally the state is changed and the update is successful. Based on this, should we introduce a bool variable to identify it?

Reply:
Any ideas?

Reply:
Actually I think the `controllerutil.OperationResult` is distinct instead of confusing. If return bool, it's more confusing from my point of view😅

Reply:
> My concern is the controllerutil.OperationResult is specific to CreateOrUpdate.

Got it. So should we create our own types here? I'm OK with the suggestion.

Reply:
Hold on for a while. I'm looking at it.

Reply:
Maybe we can pass the `` to `AggregateResourceBindingWorkStatus` which knows when and how to fire the event:
```diff
-       err = helper.AggregateResourceBindingWorkStatus(c.Client, binding, workload)                                 
+       err = helper.AggregateResourceBindingWorkStatus(c.Client, binding, workload, c.EventRecorder)
```

By the way, the two places don't need to fire events but reply on retry.
https://github.com/karmada-io/karmada/blob/10990cd96c1d1dd20ee3e2a44dbd58145f529a1b/pkg/util/helper/workstatus.go#L40-L49

Reply:
Do you mean pass event recorder to this function? I'm afraid there are more functions need to be updated too, like `ensureWork` and `ApplyPolicy`.

> By the way, the two places don't need to fire events but reply on retry.

Return `controllerutil.OperationResultNone` does not change the events behavior in this PR. When err is not nil, we still fire events in `binding_controller.go`.

```go
	operationResult, err := helper.AggregateResourceBindingWorkStatus(c.Client, binding, workload)
	if err != nil {
		klog.Errorf("Failed to aggregate workStatuses to resourceBinding(%s/%s). Error: %v.",
			binding.GetNamespace(), binding.GetName(), err)
		c.EventRecorder.Event(binding, corev1.EventTypeWarning, workv1alpha2.EventReasonAggregateStatusFailed, err.Error())
		c.EventRecorder.Event(workload, corev1.EventTypeWarning, workv1alpha2.EventReasonAggregateStatusFailed, err.Error())
		errs = append(errs, err)
	} else if operationResult != controllerutil.OperationResultNone {
		msg := fmt.Sprintf("Update resourceBinding(%s/%s) with AggregatedStatus successfully.", binding.Namespace, binding.Name)
		klog.V(4).Infof(msg)
		c.EventRecorder.Event(binding, corev1.EventTypeNormal, workv1alpha2.EventReasonAggregateStatusSucceed, msg)
		c.EventRecorder.Event(workload, corev1.EventTypeNormal, workv1alpha2.EventReasonAggregateStatusSucceed, msg)
	}
```

Reply:
It makes sense. I have made a change, left `[ensureWork](https://github.com/karmada-io/karmada/blob/2a0610eb0053dac9aa2688935fde3926c2b7db8a/pkg/controllers/binding/binding_controller.go#L116)` for refactoring in the future.